core/arch/arm/include/arm64.h

```diff
@@ -197,174 +197,80 @@
-{
-	uint64_t val;
-
-	asm volatile("mrs %[val], ttbr1_el1"
-			: [val] "=r" (val));
-	return val;
-}
-
-static inline void write_ttbr1_el1(uint64_t val)
-{
-	asm volatile("msr ttbr1_el1, %[val]"
-			: : [val] "r" (val));
-}
-
-static inline uint64_t read_mpidr_el1(void)
+static inline void isb(void)
 {
-	uint64_t val;
-
-	asm volatile("mrs %[val], mpidr_el1"
-			: [val] "=r"(val));
-	return val;
+	asm volatile ("isb");
 }
 
-static inline uint32_t read_contextidr_el1(void)
+static inline void dsb(void)
 {
-	uint32_t val;
-
-	asm volatile("mrs %[val], contextidr_el1"
-			: [val] "=r"(val));
-	return val;
+	asm volatile ("dsb sy");
 }
 
-static inline uint64_t read_cntpct(void)
-{
-	uint64_t val;
+/*
+ * Templates for register read/write functions based on mrs/msr
+ */
 
-	asm volatile("mrs %[val], cntpct_el0"
-			: [val] "=r"(val));
-	return val;
+#define DEFINE_REG_READ_FUNC_(reg, type, asmreg)	\
+static inline type read_##reg(void)			\
+{							\
+	type val;					\
+							\
+	asm volatile("mrs %0, " #asmreg : "=r" (val));	\
+	return val;					\
 }
 
-static inline uint32_t read_cntfrq(void)
-{
-	uint32_t frq;
-
-	asm volatile("mrs %[frq], cntfrq_el0"
-			: [frq] "=r"(frq));
-	return frq;
+#define DEFINE_REG_WRITE_FUNC_(reg, type, asmreg)		\
+static inline void write_##reg(type val)			\
+{								\
+	asm volatile("msr " #asmreg ", %0" : : "r" (val));	\
 }
 
-static inline uint64_t read_esr_el1(void)
-{
-	uint64_t val;
+#define DEFINE_U32_REG_READ_FUNC(reg) \
+		DEFINE_REG_READ_FUNC_(reg, uint32_t, reg)
 
-	asm volatile("mrs %[val], esr_el1"
-			: [val] "=r"(val));
-	return val;
-}
+#define DEFINE_U32_REG_WRITE_FUNC(reg) \
+		DEFINE_REG_WRITE_FUNC_(reg, uint32_t, reg)
 
-static inline uint64_t read_far_el1(void)
-{
-	uint64_t val;
+#define DEFINE_U32_REG_READWRITE_FUNCS(reg)	\
+		DEFINE_U32_REG_READ_FUNC(reg)	\
+		DEFINE_U32_REG_WRITE_FUNC(reg)
 
-	asm volatile("mrs %[val], far_el1"
-			: [val] "=r"(val));
-	return val;
-}
+#define DEFINE_U64_REG_READ_FUNC(reg) \
+		DEFINE_REG_READ_FUNC_(reg, uint64_t, reg)
 
-static inline void isb(void)
-{
-	asm volatile ("isb");
-}
+#define DEFINE_U64_REG_WRITE_FUNC(reg) \
+		DEFINE_REG_WRITE_FUNC_(reg, uint64_t, reg)
 
-static inline void dsb(void)
-{
-	asm volatile ("dsb sy");
-}
+#define DEFINE_U64_REG_READWRITE_FUNCS(reg)	\
+		DEFINE_U64_REG_READ_FUNC(reg)	\
+		DEFINE_U64_REG_WRITE_FUNC(reg)
 
-static inline uint32_t read_cpacr_el1(void)
-{
-	uint32_t val;
+/*
+ * Define register access functions
+ */
 
-	asm volatile("mrs %0, cpacr_el1" : "=r" (val));
-	return val;
-}
+DEFINE_U32_REG_READWRITE_FUNCS(cpacr_el1)
+DEFINE_U32_REG_READWRITE_FUNCS(daif)
+DEFINE_U32_REG_READWRITE_FUNCS(fpcr)
+DEFINE_U32_REG_READWRITE_FUNCS(fpsr)
 
-static inline void write_cpacr_el1(uint32_t val)
-{
-	asm volatile("msr cpacr_el1, %0" : : "r" (val));
-}
+DEFINE_U32_REG_READ_FUNC(contextidr_el1)
 
-static inline uint32_t read_fpcr(void)
-{
-	uint32_t val;
+DEFINE_REG_READ_FUNC_(cntfrq, uint32_t, cntfrq_el0)
```

Comment:
`read_cntfrq` is also implemented in `arm32.h`, where it would not make much sense to rename the function `read_cntfrq_el0` I think. Or we could do:
```
DEFINE_REG_U32_READ_FUNC(cntfrq_el0)

static inline uint32_t read_cntfrq(void)
{
    return read_cntfrq_el0();
}
```
But is it worth the trouble? Same goes for `cntpct`.